### PR TITLE
COMMON: Support : in mac filenames.

### PR DESCRIPTION
**Pull request by phcoder**

### CODE REVIEW

File: common/path.cpp [hunk 1]
```diff
@@ -178,8 +231,107 @@ Path Path::join(const String &str, char separator) const {
 
 Path Path::join(const char *str, char separator) const {
 	Path temp(*this);
-	temp.joinInPlace(str, DIR_SEPARATOR);
+	temp.joinInPlace(str, separator);
 	return temp;
 }
 
+Common::Array<Common::String> Path::splitComponents() const {
```
Comment (sev-) [inline common/path.cpp:238]: Please use `StringArray` here for brevity (`#include "common/str-array.h"`)

File: common/path.cpp [hunk 2]
```diff
@@ -178,8 +231,107 @@ Path Path::join(const String &str, char separator) const {
...
+	Common::String cur;
+	for (uint i = 0; i < _str.size(); i++) {
+		if (_str[i] == ESCAPER) {
+			i++;
+			if (_str[i] == ESCAPE_SLASH)
+				cur += '/';
+			else if (_str[i] == ESCAPE_SEPARATOR) {
+				res.push_back(cur);
+				cur = "";
+			}
```
Comment (sev-) [inline common/path.cpp:249]:
```C
} else {
    error("Path::splitComponents(): Malformed Common::Path. '%c' unexpected after '/'", _str[i]);
}
```

File: common/path.cpp [hunk 3]
```diff
@@ -178,8 +231,107 @@ Path Path::join(const String &str, char separator) const {
...
+			cur += _str[i];
+	}
+
+	res.push_back(cur);
+
+	return res;
+}
+
+Path Path::punycodeDecode() const {
+	Common::Array<Common::String> c = splitComponents();
```
Comment (sev-) [inline common/path.cpp:260]: StringArray here and everywhere else below

File: common/path.cpp [hunk 4]
```diff
@@ -20,67 +20,119 @@
...
+		else
+			ret += in[i];
+	}
+	return ret;
+}
+
+Path Path::appendComponent(const String &x) const {
+	if (x.empty())
+		return *this;
+	Common::String str = _str;
```
Comment (sev-) [inline common/path.cpp:110]: You're in namespace Common, may write just `String`

File: common/path.cpp [hunk 5]
```diff
@@ -178,8 +231,107 @@ Path Path::join(const String &str, char separator) const {
...
+		if (i + 1 < c.size())
+			res += DIR_SEPARATOR;
+	}
+
+	Path ret;
+	ret._str = res;
+	return ret;
+}
+
+static String getIdentifierComponent(const String& in) {
```
Comment (sev-) [inline common/path.cpp:274]: Could you please add a brief comment explaining what "identifier" mean here? Or is it simply "getPathComponent"?

File: common/path.cpp [hunk 6]
```diff
@@ -178,8 +231,107 @@ Path Path::join(const String &str, char separator) const {
...
+
+	for (uint i = 0; i < c.size(); i++) {
+		if (!getIdentifierComponent(c[i]).matchString(getIdentifierComponent(cpat[i]), true))
+			return false;
+	}
+
+	return true;
+}
+
+bool Path::IgnoreCaseAndMac_EqualsTo::operator()(const Path& x, const Path& y) const {
```
Comment (sev-) [inline common/path.cpp:329]: Please add a brief comment on what "Mac" mean here, e.g. explain the matching principles.

**Review by sev- — APPROVED** [review]
This is awesome!

So much cleaner, and finally, we're going away from the path separator ambiguities.

I had a few tiny notes; other than that, it looks very well.

File: common/path.cpp (hunk 3 above)
Comment (phcoder) [reply common/path.cpp:260]: Done

File: common/path.cpp (hunk 2 above)
Comment (phcoder) [reply common/path.cpp:249]: Done

File: common/path.cpp (hunk 5 above)
Comment (phcoder) [reply common/path.cpp:274]: Done

File: common/path.cpp (hunk 6 above)
Comment (phcoder) [reply common/path.cpp:329]: Done

File: common/path.cpp (hunk 4 above)
Comment (phcoder) [reply common/path.cpp:110]: Done

**Pull request merged**